CONTRIBUTING.md

```diff
@@ -0,0 +1,193 @@
+Teams that have moved swiftly at the beginning of projects can find themselves moving at a much slower rate further in. The term [Technical Debt](https://en.wikipedia.org/wiki/Technical_debt) describes the cost of the additional work that needs to be done (because of slower debugging, more bugs, etc) due to messy code.
+
+## Best Practices
+
+These best practices are mostly taken from the book `Clean Code, A Handbook of Agile Craftsmanship` by Robert C. Martin 
+
+### 1. Functions
+
+#### 1.1 Do one thing
+
+Functions should do one thing. They should do it well. They should do it only. One symptom of doing more than one thing is having sections within functions. These can oftentimes easily be split up into subroutines that will only have a specific role.
+
+#### 1.2 Size
+
+Size does matter! The size of the function should be **small**. A function that cannot be read in one screen should definitely be split up.
+
+#### 1.3 Duplication
+
+Duplicating code is one of the primary source of hard to maintain code. Here is an example of duplicated code in the code base of robotics-prototype
+
+```
+void Commands::ttoOn(void) {
+    throttleTimeOut = true;
+    String msg = "ASTRO Throttle Timeout " + String(throttleTimeOut ? "On" : "Off");
+    println(msg);
+
+}
+void Commands::ttoOff(void) {
+    throttleTimeOut = false;
+    String msg = "ASTRO Throttle Timeout " + String(throttleTimeOut ? "On" : "Off");
+    println(msg);
+
+}
+```
+
+Which can just be simplified to
+
+```
+void Commands::setTto(bool throttleTimeOut) {
+    String msg = "ASTRO Throttle Timeout " + String(throttleTimeOut ? "On" : "Off");
+    println(msg);
+
+}
+```
+
+What we gain by eliminating duplication is the need to change the function multiple times to make changes. If the logic above need to add another instruction, we would need to change `ttoOn(void)` and `ttoOff(void)` while now we only need to change `setTto(bool)`
+
+There are more changes we could do to improve this function that we will see later.
+
+
+### 2. Coupling
+Coupling is the degree of interdependence between software modules.
+
+#### 2.1 Modularity
+Since the rover needs to be modular, we need to have very loose coupling. The software needs to have independent modules that we can quickly swap out. This is one of the many reasons we use ROS, which creates a modular foundation.
+
+#### 2.2 Levels of abstraction
+It is a best practice to abstract the code into different layers. For example, code related to communication with ROS should not include GUI changes. Common layers are GUI, Logic, Communication.
+
+
+### 3. Comments
+Comments can be a blessing or a cure. Proper documentation can help developers while bad documentation can obstruct or mislead them.
+
+#### 3.1 Useless comments
+Ideal code is self explanatory. Do not comment when it is obvious like
+
+```
+//Initialize ROS
+initRos();
+```
+
+This adds lines for the programmers to read without adding any value.
+
+#### 3.2 Rotted Comments
```

Comment:
I think rotted is used as a verb, so maybe rotten (adj) would be better?

Reply:
https://wikidiff.com/rotten/rotted

The former is what I want to express. "overridden with bacteria and other infectious agents"

Reply:
Ah nevermind, I have one IQ right now. I didn't even realize it was Rotten I needed